### PR TITLE
Incremental training implementation for Drain

### PR DESCRIPTION
**Pull request by kokonguyen191**
This implementation keeps track of the trained tree and logCluL with `pickle`. Templates will keep being updated but old logs are not. There's also training history backup in case anything goes wrong.

### CODE REVIEW

File: logparser/Drain/Drain.py [hunk 1]
```diff
@@ -276,13 +391,15 @@ def parse(self, logName):
 
             count += 1
             if count % 1000 == 0 or count == len(self.df_log):
-                print('Processed {0:.1f}% of log lines.'.format(count * 100.0 / len(self.df_log)))
-
+                print('Proscessed {0:.1f}% of log lines.'.format(count * 100.0 / len(self.df_log)))
 
-        if not os.path.exists(self.savePath):
```
Comment (kokonguyen191) [inline logparser/Drain/Drain.py]: Moved to __init__

**Comment by kokonguyen191** [comment]
`HDFS_1k_1.log` and `HDFS_1k_2.log` are `HDFS_2k.log` split into 2 halves, each containing 1000 lines. Running with `resume_training=True` on `HDFS_1k_1.log` then `HDFS_1k_2.log` returns the same templates as using the original Drain on `HDFS_2k.log`.

I have also tested on my own dataset (very large, 160MB) split into 4 parts and the final template csv when run incrementally has the same SHA1 hash as the template generated by the original Drain.

**Comment by jimzhu** [comment]
Thank you for your pr! We will have a review.

**Comment by AnkitNassa** [comment]
@kokonguyen191 This looks interesting. Could you please share the history file which has the templates and builds tree using that file.

**Comment by kokonguyen191** [comment]
@AnkitNassa You can try running it with `resume_training=True` and `pickle.load` the file to see what's inside. Currently my files are a bit messy so I can't find any history file.

**Comment by jimzhu** [comment]
Thanks for the pr. To make the code clean for understanding the main algorithm, we choose to keep the current simple implementation. We will leave the engineering issues in a future version.

**Pull request closed without merging**